packages/polling-controller/src/AbstractPollingController.ts

```diff
@@ -0,0 +1,138 @@
+
+/**
+ * AbstractPollingControllerBaseMixin
+ *
+ * @param Base - The base class to mix onto.
+ * @returns The composed class.
+ */
+export function AbstractPollingControllerBaseMixin<TBase extends Constructor>(
+  Base: TBase,
+) {
+  abstract class AbstractPollingControllerBase
+    extends Base
+    implements IPollingController
+  {
+    readonly #pollingTokenSets: Map<PollingTokenSetId, Set<string>> = new Map();
+
+    #callbacks: Map<
+      PollingTokenSetId,
+      Set<(PollingTokenSetId: PollingTokenSetId) => void>
+    > = new Map();
+
+    abstract _executePoll(
+      networkClientId: NetworkClientId,
+      options: Json,
+    ): Promise<void>;
+
+    abstract _startPollingByNetworkClientId(
+      networkClientId: NetworkClientId,
+      options: Json,
+    ): void;
+
+    abstract _stopPollingByPollingTokenSetId(key: PollingTokenSetId): void;
+
+    startPollingByNetworkClientId(
+      networkClientId: NetworkClientId,
+      options: Json = {},
+    ): string {
+      const pollToken = random();
+      const key = getKey(networkClientId, options);
+      const pollingTokenSet =
+        this.#pollingTokenSets.get(key) || new Set<string>();
+      pollingTokenSet.add(pollToken);
+      this.#pollingTokenSets.set(key, pollingTokenSet);
+
+      if (pollingTokenSet.size === 1) {
+        // Start new polling only if it's the first token for this key
+        this._startPollingByNetworkClientId(networkClientId, options);
+      }
+
+      return pollToken;
+    }
+
+    stopAllPolling() {
+      this.#pollingTokenSets.forEach((tokenSet, _key) => {
+        tokenSet.forEach((token) => {
+          this.stopPollingByPollingToken(token);
+        });
+      });
+    }
+
+    stopPollingByPollingToken(pollingToken: string) {
+      if (!pollingToken) {
+        throw new Error('pollingToken required');
+      }
+
+      let keyToDelete: PollingTokenSetId | null = null;
+      for (const [key, tokenSet] of this.#pollingTokenSets) {
+        if (tokenSet.delete(pollingToken)) {
+          if (tokenSet.size === 0) {
+            keyToDelete = key;
+          }
+          break;
+        }
+      }
+
+      if (keyToDelete) {
+        // TODO figure out why this is necessary
+        const nonNullKey = keyToDelete;
+        this._stopPollingByPollingTokenSetId(nonNullKey);
+        this.#pollingTokenSets.delete(nonNullKey);
+        this.#callbacks.get(nonNullKey)?.forEach((callback) => {
+          // for some reason this typescript can't tell that keyToDelete is not null here
+          callback(nonNullKey);
+        });
+        this.#callbacks.get(nonNullKey)?.clear();
+      }
+    }
+
+    onPollingCompleteByNetworkClientId(
+      networkClientId: NetworkClientId,
+      callback: (networkClientId: NetworkClientId) => void,
+      options: Json = {},
+    ) {
+      const key = getKey(networkClientId, options);
+      const callbacks = this.#callbacks.get(key) || new Set<typeof callback>();
```

Comment:
Nit: `??` would be safer:

``` suggestion
      const callbacks = this.#callbacks.get(key) ?? new Set<typeof callback>();
```

Reply:
done here: https://github.com/MetaMask/core/pull/3636/commits/823aaa188f572a8aa76973a012ae9bfa8da518e0